src/V3Randomize.cpp

```diff
@@ -1303,6 +1303,71 @@ class RandomizeVisitor final : public VNVisitor {
+                                        AstNodeExpr* exprp) {
+        V3UniqueNames* uniqueNamep = new V3UniqueNames{"__Vrandarr"};
+        AstNodeDType* tempDTypep = dtypep;
+        AstVar* randLoopIndxp = nullptr;
+        AstNodeStmt* stmtsp = nullptr;
+
+        if (VN_CAST(tempDTypep, DynArrayDType)) {
+            AstCMethodHard* tempElementp = nullptr;
+            while (VN_CAST(tempDTypep, DynArrayDType)) {
+                AstVar* const newRandLoopIndxp
+                    = new AstVar{fl, VVarType::VAR, uniqueNamep->get(""),
+                                 dtypep->findBasicDType(VBasicDTypeKwd::UINT32)};
+                randLoopIndxp = AstNode::addNext(randLoopIndxp, newRandLoopIndxp);
+                tempElementp
+                    = tempElementp
+                          ? new AstCMethodHard{fl, tempElementp, "atWrite",
+                                               new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}}
+                          : new AstCMethodHard{fl, exprp, "atWrite",
+                                               new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}};
+                tempElementp->dtypep(tempDTypep->subDTypep());
+                tempDTypep = tempDTypep->virtRefDTypep();
+            }
+            if (VN_IS(tempElementp->dtypep()->skipRefp(), StructDType)) {
+                tempElementp->v3warn(
+                    E_UNSUPPORTED,
+                    "Unsupported: CreateArrayForeachLoop currently does not support "
+                    "this data type. (Struct-Array unconstrained "
+                    "randomization is not fully supported)");
+            }
+            AstSelLoopVars* const randLoopVarp
+                = new AstSelLoopVars{fl, exprp->cloneTree(false), randLoopIndxp};
+            stmtsp = new AstForeach{fl, randLoopVarp, newRandStmtsp(fl, tempElementp, nullptr)};
+        } else if (VN_CAST(tempDTypep, UnpackArrayDType)) {
+            AstArraySel* tempElementp = nullptr;
+            while (VN_CAST(tempDTypep, UnpackArrayDType)) {
+                AstVar* const newRandLoopIndxp
+                    = new AstVar{fl, VVarType::VAR, uniqueNamep->get(""),
+                                 dtypep->findBasicDType((VBasicDTypeKwd::UINT32))};
+                randLoopIndxp = AstNode::addNext(randLoopIndxp, newRandLoopIndxp);
+                tempElementp
+                    = tempElementp
+                          ? new AstArraySel{fl, tempElementp,
+                                            new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}}
+                          : new AstArraySel{fl, exprp,
+                                            new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}};
+                tempElementp->dtypep(tempDTypep->subDTypep());
+                tempDTypep = tempDTypep->virtRefDTypep();
+            }
+            if (VN_IS(tempElementp->dtypep()->skipRefp(), StructDType)) {
+                tempElementp->v3warn(
```

Comment:
I see from the test output the error currently points at the class, it would be more helpful if it pointed at the offending dtype in question, or the offending declaration.

Same above. (I'd also say these two code paths could be unified more, but that's already been said before.)

```suggestion
                tempElementp->dtypep()->v3warn(
```
(I also think that the checks should primarily appear on unhandled paths (like the `else` below). Proactive checking for every ill-supported corner case is hard: you can never be sure you handled all types, like maybe arrays of queues? queues of arrays? arrays of strings? unpacked arrays of dynamic arrays of unpacked arrays? this is probably never ending, so it might be simpler for the support to be generic and for bugs to be found later. I don't want to say *worse is better*, I don't mind the check, to me it can stay as it is, just wanted to give some feedback.)

Reply:
Thank you for your suggestions and feedback, and I completely agree with you. The current check is primarily meant to inform users that this feature is not yet supported. I believe that this issue will be resolved in a few days or weeks, and once it's addressed, this temporary check will be removed.

I believe the current handling is fairly generic, but there are indeed issues with struct types, which I have mentioned in the related issue #5455 . My plan is that once support for other array types is in place, a more comprehensive test will definitely be necessary. I'll make sure to test most scenarios to ensure the stability of the code. Thanks again for your valuable input!